ipn/ipnlocal/network-lock.go

```diff
@@ -96,12 +96,92 @@ func (b *LocalBackend) tkaSyncIfNeededLocked(nm *netmap.NetworkMap) error {
 	}
 
 	if isEnabled && b.tka.authority.Head() != nm.TKAHead {
-		// TODO(tom): Implement sync
+		if err := b.tkaSyncLocked(ourNodeKey); err != nil {
+			return fmt.Errorf("tka sync: %v", err)
+		}
 	}
 
 	return nil
 }
 
+func toSyncOffer(head string, ancestors []string) (tka.SyncOffer, error) {
+	var out tka.SyncOffer
+	if err := out.Head.UnmarshalText([]byte(head)); err != nil {
+		return tka.SyncOffer{}, fmt.Errorf("head.UnmarshalText: %v", err)
+	}
+	out.Ancestors = make([]tka.AUMHash, len(ancestors))
+	for i, a := range ancestors {
+		if err := out.Ancestors[i].UnmarshalText([]byte(a)); err != nil {
+			return tka.SyncOffer{}, fmt.Errorf("ancestor[%d].UnmarshalText: %v", i, err)
+		}
+	}
+	return out, nil
+}
+
+// tkaSyncLocked synchronizes TKA state with control. b.mu must be held
```

Comment:
I feel like it might be nice to do this as a non-Locked receiver, only grabbing the mutex for the small moments when needed. There's nothing wrong with this approach, it just feels like it would be better to hold the lock for the minimum time required. It doesn't really change anything if you do unmarshalling while the lock is released except that without RAII, the lock management is more annoying that way.

Reply:
I'm not sure about inverting it because `b.tka` needs the lock and its used in about 4 places, and we spend a lot of time computing stuff against `b.tka.authority` and `b.tka.storage`, so really the only times it safe to release the lock are the network RPCs.